### DISCUSSION TITLE
Make Search "Full Text"

### DISCUSSION THREAD

**Discussion by doodlemania2** [directus/directus#4492] [Feature Requests]
In the search bar: 
![image](https://user-images.githubusercontent.com/5252217/110837949-81736d80-8267-11eb-88b1-6e6121581396.png)

Would be great if we could make that search bar full text, with support for wildcards. Examples:

1. Application Gateway --> Cannot find if I search the word Gateway.
2. Topical Whatever Hootis --> Cannot find if I search the word Hootis.

In the above examples, I have to start by typing Application or Topical.

**Comment by rijkvanzanten** [comment]
It's currently using LIKE with the value surrounded with `%{value}%`. I always though that'd do the trick in SQL 🤔

**Reply by hostmaster-wikisoft** [reply]
Using postgres you can speed up the search many many many times using full text search tsvectors, GIN indexes and they even have a websearch_to_tsquery which supports and's and or's. You get pluralis, stopwords, ranking etc. it's basically a pretty good search engine, using very simple tricks, and works with offsetting and limits for paging, and combined where clauses for filtering/faceted search. And you don't have to sync your data to another system, but simply make a autogenerated column with the search vectors.

Using this approach I can easily search through 57 mio rows, in sub 100ms (actually when it hits the index correct it's more like sub 20ms, even for and and or's). Also ran a couple of test on 340 mio rows and it works pretty slick. Specially if you partition the collection table.

You'll never use like '%[query]%' again.

https://rob.conery.io/2019/10/29/fine-tuning-full-text-search-with-postgresql-12/

I know this is rather postgres specific, but if you could let us hack into the search functionality and return our own sql queries og let us set a search endpoint like the autocomplete interface (under settings maybe), with som information about the search, collection etc and a schema for returning the results. It would be possible to both enhance the search in the interface with speed, but maybe also with external data (virtual columns pr item or even virtual items) from other systems?

**Reply by benhaynes** [reply]
I wonder if there's a way for us to standardize the vendor-specific enhancements through some sort of extension code. @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
@benhaynes Certainly. I was already dreaming up an extension type that would allow you to connect to algolia/elasticsearch thought the `search` query param. That being said, it's a pretty in-depth new feature, not something we can just easily tack on

**Reply by raphaeltm** [reply]
@rijkvanzanten  I assume this is way out of scope for the current efforts at Directus, but https://www.meilisearch.com/ is another full-text search engine that is open source and much lighter-weight/simpler than elasticsearch. They've been doing amazing work on it lately. Could be a neat addition if you eventually do go down the full-text search path 😊

**Reply by rijkvanzanten** [reply]
@raphaeltm I would love to have a modular setup for more advanced full text search, that would allow to configure the use of "local" (current/default) or a provider like "meilisearch" or "elasticsearch". That way the sky is the limit (and it'll match neatly with how we're solving file storage / auth / etc etc

**Reply by programmarchy** [reply]
To piggyback on @hostmaster-wikisoft being able to tokenize search would open up a lot of power, especially in Postgres.

I would add that it's not quite Postgres specific, because MySQL supports FTS natively, and SQLite has a FTS extension.

Search bar would be nice, but even better would be availability in the Directus API.

One way to add it would be to add a new filter operator similar to `_contains`/`_icontains` which used FTS instead of LIKE. Maybe the operator could be named something like `_search` which could tap into FTE if supported by the storage provider, otherwise fall back to the `_icontains` implementation.

Another building block for the GraphQL API might be a concat() function. So instead of something like:

```graphql
query {
  my_things(filter: {
    _or: [ title: { _search: "yee haw" }, description: { _search: "yee haw"} ]
  }) {
    title
    description
  }
}
```

You could do:

```graphql
query {
  my_things(filter: {
    concat(fields: { title, description }) { _search: "yee haw" }
  }) {
    title
    description
  }
}
```

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
